File: BITEC/lessons/block_2_object_orientation/src/lerneinheiten/L02KlassenUndMethoden/grundlagen/Mensch.java
package lerneinheiten.L02KlassenUndMethoden.grundlagen;

public class Mensch {
    // Attribute sind nomen
    private String name;
    private int alter;
    private double happiness;

    // hat-Beziehungen:
    // Die Beziehung "loveInterest" (geliebte/r) muss nicht gegenseitig sein (mutual).
    // Deshalb können viele eine Person als love interest haben und auch viele keinen haben welche love interested ist.
    private Mensch loveInterest;

    // Konstruktor
    public Mensch(String name, double happiness, int alter) {
        this.name = name;
        this.happiness = happiness;
        this.alter = alter;
    }

    public Mensch(String name, double happiness, int alter, Mensch loveInterest) {
        this(name, happiness, alter);
        this.loveInterest = loveInterest;
    }

    //Methoden sind verben
    public HundeBesitzer wirdEinHundeBesitzer(Hund hund, boolean hatHundeFuehrerschein, int capacity) {
        HundeBesitzer einGanzNeuerMensch = new HundeBesitzer(this, hatHundeFuehrerschein, capacity);
        einGanzNeuerMensch.kaufen(hund);

        return einGanzNeuerMensch;
    }

    public HundeBesitzer mehrereHundeKaufen(Hund[] hunde, boolean hatHundeFuehrerschein, int capacity) {
        HundeBesitzer einGanzNeuerMensch = new HundeBesitzer(this, hatHundeFuehrerschein, capacity);

        if (capacity <= hunde.length) {
            System.out.println("Fehler! Wir haben zu viele Hunde als wir betreuen können.");
            return null; // Achtung! Wir werden ein return null in Zukunft mit Exceptions ersetzen.
        }

        for (Hund hund : hunde) {
            einGanzNeuerMensch.kaufen(hund);
        }

        return einGanzNeuerMensch;
    }

    // get und set methoden
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getHappiness() {
        return happiness;
    }

    public void setHappiness(double happiness) {
        this.happiness = happiness;
    }

    public int getAlter() {
        return alter;
    }

    public void setAlter(int alter) {
        this.alter = alter;
    }

    public Mensch getLoveInterest() {
        return loveInterest;
    }

    public void setLoveInterest(Mensch loveInterest) {
        this.loveInterest = loveInterest;
    }
}

File: JET/modul_1_grundlagen/L05KlassenMethoden/Exercise1-HundeUndBesitzer/loesung/Mensch.cs
namespace Hunde;

public class Mensch
{
    // private Felder
    private string name;
    private int alter;
    private double happiness;

    // hat-Beziehungen:
    private Mensch myLoveInterest;

    // Konstruktor
    public Mensch(string name, double happiness, int alter)
    {
        this.name = name;
        this.happiness = happiness;
        this.alter = alter;
    }

    public Mensch(string name, double happiness, int alter, Mensch loveInterest) : this(name, happiness, alter)
    {
        this.myLoveInterest = loveInterest;
    }

    //Methoden
    public HundeBesitzer WirdEinHundeBesitzer(Hund hund, bool hatHundeFuehrerschein, int capacity)
    {
        HundeBesitzer einGanzNeuerMensch = new HundeBesitzer(this, hatHundeFuehrerschein, capacity);
        einGanzNeuerMensch.Kaufen(hund);
        return einGanzNeuerMensch;
    }

    public HundeBesitzer MehrereHundeKaufen(Hund[] hunde, bool hatHundeFuehrerschein, int capacity)
    {
        HundeBesitzer einGanzNeuerMensch = new HundeBesitzer(this, hatHundeFuehrerschein, capacity);

        if (capacity <= hunde.Length)
        {
            Console.WriteLine("Fehler! Wir haben zu viele Hunde als wir betreuen können.");
            return null;
        }

        foreach (Hund hund in hunde)
        {
            einGanzNeuerMensch.Kaufen(hund);
        }

        return einGanzNeuerMensch;
    }

    public bool DetectLoveTriangle()
    {
        bool triangle = myLoveInterest.myLoveInterest.myLoveInterest == this;
        bool selfLove = myLoveInterest != this;

        return triangle && !selfLove;
    }

    private bool DetectLoveTriangleOfSize(int n)
    {
        if (myLoveInterest != this) 
            return false; 

        Mensch next = this;

        for (int i = 0; i < n; i++)
        {
            next = next.myLoveInterest;
            
        }

        return next == this;
    }

    public (bool found, int foundAtSize) DetectLoveTriangleUntilSize(int n)
    {


        for (int i = 0; i < n; i++)
        {
            if(DetectLoveTriangleOfSize(i))
            {
                return (true, i);
            }
            
        }

        return (false, -1);
    }

    public bool DetectMutualLove()
    {
        return this == myLoveInterest.myLoveInterest;
    }

    // Get-und-Set-Methoden
    public string GetName()
    {
        return name;
    }

    public void SetName(string name)
    {
        this.name = name;
    }

    public double GetHappiness()
    {
        return happiness;
    }

    public void SetHappiness(double happiness)
    {
        this.happiness = happiness;
    }

    public int GetAlter()
    {
        return alter;
    }

    public void SetAlter(int alter)
    {
        this.alter = alter;
    }

    public Mensch GetLoveInterest()
    {
        return myLoveInterest;
    }

    public void SetLoveInterest(Mensch loveInterest)
    {
        this.myLoveInterest = loveInterest;
    }
}
